### PR TITLE
fix(consensus): using WAL heights

### PR DESCRIPTION
**Pull request by vbar**
Starting new heights for pathfinder consensus after node restart happens too often:
a) heights started before the restart are restarted (which probably doesn't play well with malachite internal state)
and b) heights finalized before the restart are also restarted, upon receiving old messages from other nodes (which apparently leads to the restarted node getting stuck)

This PR makes the calls to `start_height` more selective, leading to subjectively better success of integration tests (primarily `fail_on_proposal_committed`).

### CODE REVIEW

**Comment by vbar** [comment]
> > subjectively better success
> 
> This isn't a meaningful metric.
> 
I'd like to use a better metric, but I don't have one.

> I'm fine returning finalized heights when loading from WAL. That part makes sense.
> 
> However, `finalized_heights` should **not** be kept as part of the `Consensus` engine. The `HashSet` should just be returned to the caller.
> 
https://github.com/eqlabs/pathfinder/pull/3096/commits/bbe241cf44527d6cef92f3463da6a21d0a73fc61

File: crates/consensus/src/lib.rs [hunk 1]
```diff
@@ -718,9 +718,9 @@ impl<
         }
     }
 
-    /// Get the current maximum height being tracked by the consensus engine.
-    pub fn current_height(&self) -> Option<u64> {
-        self.internal.keys().max().copied()
+    /// Get the heights currently being tracked by the consensus engine.
+    pub fn incomplete_heights(&self) -> HashSet<u64> {
+        self.internal.keys().copied().collect()
```
Comment (t00ts) [inline crates/consensus/src/lib.rs:723]: The original `current_height` method should be brought back untouched. I don't see any reason for this breaking change here.

File: crates/pathfinder/src/consensus/inner/consensus_task.rs [hunk 2]
```diff
@@ -88,32 +88,38 @@ pub fn spawn(
...
-        let mut started_heights = HashSet::new();
+        let mut started_heights = consensus.incomplete_heights();
+
+        // Get the current height
+        let mut current_height = started_heights.iter().copied().max().unwrap_or_else(|| {
+            match finalized_heights.iter().max() {
+                Some(m) => m + 1,
+                None => 0,
+            }
+        });
```
Comment (t00ts) [inline crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: This is still possible while keeping the original `Consensus::current_height()` method:

```suggestion
        let current_height = consensus
            .current_height()
            .unwrap_or_else(|| finalized_heights.iter().copied().max().map_or(0, |m| m + 1));

```

File: crates/consensus/src/lib.rs (hunk 1 above)
Comment (vbar) [reply crates/consensus/src/lib.rs:723]: The original implementation is incorrect: when there are no incomplete heights (which is a totally valid state after restart), the current height is not necessarily 0.

Comment (t00ts) [reply crates/consensus/src/lib.rs:723]: Please help me understand. AFAIU, if there are no heights, it will return `None`, not `0`.

Comment (vbar) [reply crates/consensus/src/lib.rs:723]: Hmm, yes, the conversion was done by the caller - but it's still an odd interface, failing to provide current height just because all heights are momentarily finalized... Are you planning to use it?

File: crates/pathfinder/src/consensus/inner/consensus_task.rs (hunk 2 above)
Comment (vbar) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: Possible, yes, but I don't think it's very cohesive to split the retrieval of current height (which needs both finalized and incomplete heights) between crates...

File: crates/consensus/src/lib.rs (hunk 1 above)
Comment (t00ts) [reply crates/consensus/src/lib.rs:723]: > failing to provide current height just because all heights are momentarily finalized

Interesting point! This is a different story though, so I'd say please file an issue for this one and assign it to me. I'll look into it straight away and find a good solution.

In the meantime, I say we revert this change. This crate is public, we should not arbitrarily nuke an existing method and replace it with another one when there's no real need to do so.

File: crates/pathfinder/src/consensus/inner/consensus_task.rs (hunk 2 above)
Comment (t00ts) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: I'm sorry, I don't understand.

File: crates/consensus/src/lib.rs (hunk 1 above)
Comment (vbar) [reply crates/consensus/src/lib.rs:723]: But it isn't arbitrary - it's part of the bugfix... The API is error-prone - as demonstrated by the fact the caller used it erroneously - and fixing that leaves the API without any callers at all. Of course it should be removed... As for the change being "breaking", so is changing the return type of `Consensus::recover`, and a lot of other changes besides - the whole consensus functionality is highly experimental...

File: crates/pathfinder/src/consensus/inner/consensus_task.rs (hunk 2 above)
Comment (vbar) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: As in the code above: current height is normally computed as a maximum of incomplete heights, but that cannot be done when there _are_ no incomplete heights (which is a valid if somewhat uncommon state). In that case, current height should be computed as the maximum of finalized heights plus one (and obviously that cannot be done when _are_ no finalized heights either, in which case the current height is 0). So, for reliably computing current height, the context should have both finalized and incomplete heights. Now I'm coming to think maybe the context _should_ be `Consensus`, but that's clearly a bigger change, which I'd argue is outside the scope of this PR...

File: crates/consensus/src/lib.rs (hunk 1 above)
Comment (t00ts) [reply crates/consensus/src/lib.rs:723]: The "bug" can be fixed without touching this method as described below...

We can open another issue to "fix" this method, I'm happy to bounce ideas on that regard.

File: crates/pathfinder/src/consensus/inner/consensus_task.rs (hunk 2 above)
Comment (CHr15F0x) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: My 3 cents:
1. WAL does not reliably give the current height on restore (due to it not storing the proposal payloads and intermediate execution results), hence we pass `highest_finalized` to `recover()`,
2. `current_height` should then be whatever `highest_finalized` + 1, if the consensus DB does not happen to contain a full proposal that unluckily hasn't been committed yet, even though we know it's been decided upon,
3. so in the former case I'd expect `current_height` to be `highest_finalized + 1`, whereas in the latter it should be (??) `highest_finalized + 2 == decided_proposal_height_in_DB2 + 1`.

We're clearly missing 3. rn.

BTW Personally I don't care about API breaking changes as long as the API is consistent for the caller. The feature is still very much experimental and the aim is to push with the implementation to make sure we _functionally plug all the holes_ and have robust tests around the implementation to be sure we're not missing any peculiar corner cases. In such state we can then stress test and make the implementation even more robust, with maybe some refactors on the horizon.

Comment (t00ts) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: Let's track this: https://github.com/eqlabs/pathfinder/issues/3099

@CHr15F0x could you please clarify that 3rd point? Not sure I've fully grasped it. 

Feel free to append to the issue above.

File: crates/consensus/src/lib.rs (hunk 1 above)
Comment (t00ts) [reply crates/consensus/src/lib.rs:723]: > We can open another issue to "fix" this method 

https://github.com/eqlabs/pathfinder/issues/3099

File: crates/pathfinder/src/consensus/inner/consensus_task.rs (hunk 2 above)
Comment (CHr15F0x) [reply crates/pathfinder/src/consensus/inner/consensus_task.rs:122]: Ok, I had to do some code surfing...

So let's assume, that the entire proposal has been executed and stored and even decided upon but the process is killed at [this point in the code](https://github.com/eqlabs/pathfinder/blob/cdb47e97f28d0743a2a6510bd154ce3813051463/crates/pathfinder/src/consensus/inner/consensus_task.rs#L244). 

IIUC, looking at the code, we don't have full recovery implemented in this case. Normally our node should recover at this H that was decided, by that I mean: not only recover from WAL and from the prod-DB by looking at the highest committed block at H-1, but also it should (if previously deferred, finalize H), and commit it to the prod-DB, and purge the proposal messages for H. Only after this is done, we can always assume point 2. to be true. And point 3. should then be removed from the list.
